### PR TITLE
Add sources to effects

### LINKED ISSUE #42367: Zombie blood loss deaths caused by your weapons don't count as kills by you
I noticed a zombie child bleeding out while in combat with me didn't give me a morale hit. I tested letting fat zombies bleed out from wounds caused by my spear, and when they bled out, they didn't increment the "Fat zombie" kill counter.

Thus the bug doesn't count blood loss kills as kills.

1. Attack zombie until near death.
2. Wait for blood loss to take its course.

Expected behavior

Zombies with blood loss caused by you should count as kills by you. Perhaps if you wounded them in the last 5 minutes, or if you got them down a certain percentage, it would count as your kill.

Versions and configuration

- OS: Windows
    - OS Version: 10.0 1909
- Game Version: 0.E-4329-ga3e5f32 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]
   Running Windows 10, 64 bit.
   Version is 0.E-4329-ga3e5f32, using Tiles version, English, only using DDA and Disable NPC Needs.

### PR DESCRIPTION
**Pull request by irwiss**
#### Summary

SUMMARY: None

#### Purpose of change

Fixes #42367 

Tried taking a stab at #42367 - keeping track of bleeding to attribute kills
With jbytheway's help this seems like something workable.

#### Describe the solution

This patch adds effect_source, it has a few fields but currently keeps track
of the player, npc characters or monster faction.

At this time only player field is used to attribute bleed kills - but this can
be expanded to make npcs angry at you lighting them on fire, fix harming/killing
with traps without losing pacifist, spell effects that can have a persistent target
to act upon etc.

There's around 450 calls to various add_effect overloads so to keep this
PR light I added shims to forward `effect_source::empty()` instead, if this PR
is accepted there's future work to go over the sources and assign the effect
blame and make sources propagate

#### Describe alternatives you've considered

* Having non-persistent pointer to last attacker on Creature for kill attribution with some
    hostile actions setting it and call it a day, won't be too reliable
* Slightly breaking the style of damage functions and adding source as optional parameter
    and fix just the obvious callsites with the rest going to default to nullptr

#### Testing

Apply patch, cut a monster witout killing and wait so it dies from bleeding - should now
be credited in kill counts/achievements etc, use a bow or bullets to do the same - should
also work

There's also automated test in effects_test.cpp that tests whether bleed effect get applied
properly between a few entities

### CODE REVIEW

**Comment by jbytheway** [comment]
My instinct would be that we should have an `effect_source` struct which specifies an optional `character_id` and a `faction_id`.  For creatures we can just provide the faction.  Then, in the future if we decide we need more details (like the exact creature) then the struct can be changed without having to change all the places that pass this object around.

**Comment by jbytheway** [comment]
...eventually `effect_source` might see uses beyond just `effect` object, e.g. tracking the cause of fields.  So if you can think of a name that doesn't make it seem like it's necessarily tied to effects that might be better.

**Comment by irwiss** [comment]
That makes sense

Is https://github.com/CleverRaven/Cataclysm-DDA/pull/44567/commits/be7aeb34442d38def9c8a83f34b4209522ae386b#diff-d4b7e44a8d69f9917b2dd451db6870e9

Along the lines of what you meant?

File: src/creature.cpp [hunk 1]
```diff
@@ -1108,10 +1108,11 @@ void Creature::add_effect( const efftype_id &eff_id, const time_duration &dur, b
         }
     }
 }
-void Creature::add_effect( const efftype_id &eff_id, const time_duration &dur, bool permanent,
+void Creature::add_effect( const Creature *source, const efftype_id &eff_id,
```
Comment (jbytheway) [inline src/creature.cpp:1111]: This should probably take an `effect_source` rather than a `Creature*`.

File: src/effect.h [hunk 2]
```diff
@@ -296,6 +300,11 @@ class effect
             return eff_type->id;
         }
 
+        /** Returns the creature that caused the effect. */
+        const effect_source &get_source() const;
+        /** Sets the creature that caused the effect. */
+        void set_source( const effect_source &source );
```
Comment (jbytheway) [inline src/effect.h:306]: It doesn't look like you're ever calling this function.  Better not to have it if its unused; I feel like it should never be necessary to change the source of an existing effect.

File: src/monster.cpp [hunk 3]
```diff
@@ -2393,6 +2393,17 @@ void monster::drop_items_on_death()
...
+    const auto &eff_source = it.get_source().get_character_id();
+    Creature *source_creature = nullptr;
+    if( !eff_source ) {
+        source_creature = nullptr;
+    } else if( eff_source.value() == get_player_character().getID() ) {
+        source_creature = &get_player_character();
+    } else {
+        source_creature = g->find_npc( eff_source.value() );
+    }
+
```
Comment (jbytheway) [inline src/monster.cpp:2406]: This logic should probably be encapsulated in a member function of `effect_source`.

File: src/monster.cpp [hunk 4]
```diff
@@ -2412,7 +2423,7 @@ void monster::process_one_effect( effect &it, bool is_new )
     int val = get_effect( "HURT", reduced );
     if( val > 0 ) {
         if( is_new || it.activated( calendar::turn, "HURT", val, reduced, 1 ) ) {
-            apply_damage( nullptr, bodypart_id( "torso" ), val );
+            apply_damage( source_creature, bodypart_id( "torso" ), val );
```
Comment (jbytheway) [inline src/monster.cpp:2426]: I think `apply_damage` should probably change to taking an `effect_source` argument, but that probably shouldn't be part of this PR.

**Review by jbytheway — COMMENTED** [review]
Yep, that's along the lines I was thinking.

It would be good to see tests added for this new functionality too.

File: src/savegame_json.cpp [hunk 5]
```diff
@@ -417,6 +418,40 @@ void character_id::deserialize( JsonIn &jsin )
...
+    const int char_id = data.get_int( "character_id", -1 );
+    const std::string fac_id = data.get_string( "faction_id", "" );
+    const std::string mfac_id = data.get_string( "mfaction_id", "" );
+    this->character = ( char_id == -1 )
+                      ? cata::optional<character_id>() : cata::optional<character_id>( char_id );
+    this->fac = fac_id.empty()
+                ? cata::optional<faction_id>() : cata::optional<faction_id>( fac_id );
+    this->mfac = mfac_id.empty()
+                 ? cata::optional<mfaction_id>() : cata::optional<mfaction_id>( mfaction_str_id( mfac_id ).id() );
+    // mon isn't serialized
```
[the remark this answers is not in the document]
Comment (irwiss) [reply src/savegame_json.cpp]: Right, was fighting the compiler at the time and missed this,
I think it'd make sense to save mfaction as it's string id,
is there something better to do that than what's at
https://github.com/CleverRaven/Cataclysm-DDA/pull/44567/files#diff-e9234e34de70fd8a9df663074912dbb6R440 ?

File: src/creature.cpp [hunk 6]
```diff
@@ -982,15 +982,15 @@ void Creature::set_fake( const bool fake_value )
     fake = fake_value;
 }
 
-void Creature::add_effect( const effect &eff, bool force, bool deferred )
+void Creature::add_effect( const Creature *source, const effect &eff, bool force, bool deferred )
```
[the remark this answers is not in the document]
Comment (irwiss) [reply src/creature.cpp]: Seems it's only used in 1 place and just hides what's going
on - I've put a PR to remove it, along with a change for to
make test work in

https://github.com/CleverRaven/Cataclysm-DDA/pull/44623

I've been told you could also weigh in on whether it's safe to force
fresh id on clear_character too so please do if you have the time.

Could hack around both issues but I think it'd be cleaner without
them
